src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,111 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int MINIMAL_ARRAY_INDEX = 0;
+    private static final double GROW_COEFFICIENT = 1.5;
+
+    private static final String INDEX_OUT_OF_BOUNDS_MESSAGE = "Index %d out of bounds for size %d";
+    private static final String NO_SUCH_ELEMENT_MESSAGE = "There is no such element: ";
+
+    private Object[] data = new Object[DEFAULT_CAPACITY];
+    private int size;
+
     @Override
     public void add(T value) {
-
+        if (size == data.length) {
+            data = grow();
+        }
+        data[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAddOperation(index);
+        if (size == data.length) {
+            data = grow();
+        }
 
+        System.arraycopy(data, index, data, index + 1, size - index);
+        data[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexForGetOperation(index);
+        return (T) data[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndexForGetOperation(index);
+        data[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexForGetOperation(index);
+        T removedObject = get(index);
```

Comment:
Probably we can get rid of `checkIndexForGetOperation` in this method, as it is already present in `get(index)`.